Vendoring libnetwork for the pluggable IPAM driver support

mavenugo opened the pull request:
This commit brings the backend support for 
- Pluggable IP Address Management
- Network UX changes to include ipam params
- Performance and Scalability enhancements
- Libkv update with backend store fixes (including TLS)
- Fixes a few issues in libnetwork identified in #16645 
  - https://github.com/docker/docker/issues/16887
  - https://github.com/docker/docker/issues/16909
  - https://github.com/docker/docker/pull/16645#issuecomment-144065134
  - https://github.com/docker/docker/pull/16645#issuecomment-145595087

All UT and IT for these backend features are in libnetwork project.
integration-cli updated with UX and API changes in docker.
(moved docker_cli_network_test.go to be !windows due to the usage of plugins).

mavenugo committed:
Vendoring in libnetwork ipam support

* Pluggable IPAM driver support
* Performance and Scalability enhancements
* Libkv fixes in multiple backend stores

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
IPAM API & UX

introduced --subnet, --ip-range and --gateway options in docker network
command. Also, user can allocate driver specific ip-address if any using
the --aux-address option.
Supports multiple subnets per network and also sharing ip range
across networks if the network-driver and ipam-driver supports it.
Example, Bridge driver doesnt support sharing same ip range across
networks.

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo committed:
Docker side changes for the newly introduced IPAM driver

* Made use of IPAM driver primitives for legacy IP configurations
* Replaced custom Generics with backend labels

Signed-off-by: Madhu Venugopal <madhu@docker.com>

mavenugo commented:
ping @icecrime.

I will push a follow up PR today with the UX options for ipam in `docker network create` command.

mavenugo commented:
(first) rebased on top of user-namespace support  :runner:

mavenugo commented:
Added ux and API for the ipam configurations. with corresponding integration-cli tests.

mavenugo commented:
The ipam options in the UX demands some design explanation.
These options are exactly the same that we already have in docker daemon.

`--fixed-cidr` is the `--ip-range`
`--bip` is the `--gateway`
`--bip`'s subnet is the `--subnet`

Prior to the new network UX, docker had just 1 network and that is the bridge network and the only way to configure it is via the daemon flags as mentioned above.

Now that we have multiple networks backed by different plugins, we generalized it in the IPAM configuration via the above options. For backward compatibility reasons we mapped the older daemon options to the new IPAM configurations for docker0 bridge and hence the existing configurations are not impacted.

For the new networks created by `docker network create`, they all get the same functionality and more, based on the backend IPAM driver and Network Driver.

nerdalert commented:
Tested network creates/subnet/connects along with loading external driver default IPAM `Join()` calls and they all worked as advertised  :+1:

tomdee commented:
What's the plan for documentation?

mavenugo commented:
@tomdee the networking docs are being completely overhauled as part of https://github.com/moxiegirl/docker/tree/new-networking-docs. No point adding it now and changing it shortly after.

But, we are adding developer docs to the remote driver APIs in libnetwork project just like we have for remote network driver. thanks for opening https://github.com/docker/libnetwork/issues/635

tomdee commented:
Cool, so IPAM will be covered in https://github.com/moxiegirl/docker/tree/new-networking-docs too? That sounds like a good plan.

I'm having some issues with this branch when running with a remote driver.

I'm starting the daemon with
```
sudo ip link set docker0 down && sudo brctl delbr docker0 && sudo ip link set docker_gwbridge down && sudo brctl delbr docker_gwbridge ;sudo docker-dev daemon -D
```

Then I try to create a network and it fails
```
$ docker-dev network create -d calico test
Error response from daemon: Invalid Address Pool
```

The daemon outputs
```
DEBU[0024] Calling POST /v1.21/networks/create          
INFO[0024] POST /v1.21/networks/create                  
DEBU[0024] calico's manifest: &{[NetworkDriver]}        
DEBU[0024] calico implements: NetworkDriver             
DEBU[0024] Allocating IPv4 pools for network test (fa4e29153232e1c9b4d9b95a2bfed93c909c1885b3b1d02f575b9aa807f17a1f) 
DEBU[0024] RequestPool(GlobalDefault, , , map[], false) 
ERRO[0024] Handler for POST /v1.21/networks/create returned error: Invalid Address Pool 
ERRO[0024] HTTP Error                                    err=Invalid Address Pool statusCode=500
```

Let me know if you'd like me to raise this as a separate issue.

mavenugo commented:
@tomdee yes. the docs will be part of that effort.

The error message is originated from the calico driver. it is not a libnetwork issue.
We can chat in IRC and I can explain the network-driver role on how to make use of the IPAM data.

tomdee commented:
I've worked with @madhu on IRC and identified a few issues here.

The problem I was hitting had nothing to do with the Calico driver. I could repro them all with the overlay driver.

1. Using a global driver (such as Calico or the Overlay driver) can't be used with Docker unless the daemon is started with the --cluster-store flag. The current error message (that I raised above `Error response from daemon: Invalid Address Pool`) is no good. It needs to tell the user that they need to start the Docker daemon with the `--cluster-store` flag since they are using a global driver.
2. Passing in subnet and ip-range to work around the problem `docker-dev network create --subnet=10.10.0.0/16 --ip-range=10.10.10.0/24 -d overlay my_overlay` causes docker to crash. Output is here https://www.irccloud.com/pastebin/JnugtTo3/
3. Using --cluster-store with etcd fails. `ERRO[0012] Handler for POST /v1.21/networks/create returned error: failed to update store for object type *libnetwork.network: 104: Not a directory (/docker/network/v1.0/network) [10] `

I'm going to continue testing using --cluster-store with `consul`.

vieux commented:
The naming issue seems to be resolved with this PR. The swarm CI will be green again :)
LGTM for this matter.

mavenugo commented:
@tomdee yes. drivers that are declared as global drivers require `--cluster-store` by design. So, it is not a bug, but an pre-req. 

Yes, the error message (`Invalid Address Pool`) can be better and we can fix that in the next vendor-in.

abronan commented:
LGTM for everything related to the storage.

mavenugo commented:
@tiborvass 
> what is --aux-address for ?

Aux-address are auxilary-addresses that users might want to configure for some network drivers that need it to be user configured and make sure to be allocated by ipam-driver. This is because we dont want the containers to be allocating that IP that network-driver might want to use for other purposes (say for example some driver might need a firewall-ip or router-ip, etc... These are beyond the standard --subnet, --gateway and --fixed-cidr configurations that is used in typical ipam.
So --aux-address provides the flexibility that the operators would need if they are using a specific network-driver that has some special capability that is not as standard such as a gateway.

thaJeztah commented:
@mavenugo based on the above, this includes UX changes in docker as well? Will there be docs as part of this pr?

mavenugo commented:
@thaJeztah the docs are all updated as part of https://github.com/moxiegirl/docker/tree/new-networking-docs. If you prefer docs to be updated here, then we may have some wasted effort here as the docs overhaul for networking is big which @moxiegirl is working with @dave-tucker.

thaJeztah commented:
Thanks @mavenugo, I wasn't sure if these changes were part of that work as well

icecrime commented:
Can we update the top comment to add `Fixes #...`, because we have many opened issues on the PR and little idea which will get closed by this.

tomdee commented:
Another problem I've spotted during testing. Again, let me know if you want a separate issue for it

Docker inspect output shows  `Gateway": "\u003cnil\u003e",`

tomdee commented:
And another to do with how /etc/hosts is filled out.

I have two containers on the `calico` network. `tom1` has lost the etc/hosts entry for `tom2`. But `tom2` has an entry for `tom1`

```
gulfstream@gulfstream:~/calico-docker$ docker exec tom1 cat /etc/hosts
10.0.3.2	9ae211bd8573
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.23.0.3	gateway_b60623767ecf
172.23.0.3	gateway_b60623767ecf.docker_gwbridge
gulfstream@gulfstream:~/calico-docker$ docker exec tom2 cat /etc/hosts
10.0.3.3	b60623767ecf
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
10.0.3.2	tom1
10.0.3.2	tom1.calico
172.23.0.2	gateway_9ae211bd8573
172.23.0.2	gateway_9ae211bd8573.docker_gwbridge
```
Now if I run another container on that network, it gets added to both `tom1` and `tom2` but it's added twice to both.
```
gulfstream@gulfstream:~/calico-docker$ docker exec tom2 cat /etc/hosts
10.0.3.3	b60623767ecf
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
10.0.3.2	tom1
10.0.3.2	tom1.calico
172.23.0.2	gateway_9ae211bd8573
172.23.0.2	gateway_9ae211bd8573.docker_gwbridge
10.0.3.4	mad_euclid
10.0.3.4	mad_euclid.calico
10.0.3.4	mad_euclid
10.0.3.4	mad_euclid.calico
172.23.0.4	gateway_5ce47e8bb588
172.23.0.4	gateway_5ce47e8bb588.docker_gwbridge
gulfstream@gulfstream:~/calico-docker$ docker exec tom1 cat /etc/hosts
10.0.3.2	9ae211bd8573
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.23.0.3	gateway_b60623767ecf
172.23.0.3	gateway_b60623767ecf.docker_gwbridge
10.0.3.4	mad_euclid
10.0.3.4	mad_euclid.calico
10.0.3.4	mad_euclid
10.0.3.4	mad_euclid.calico
172.23.0.4	gateway_5ce47e8bb588
172.23.0.4	gateway_5ce47e8bb588.docker_gwbridge
```

mavenugo commented:
@tomdee will add omitempty for Gateway and that will solve the encoding problem.

the /etc/hosts issue is not addressed in this PR. I think there is an issue open for that.

mavenugo commented:
@icecrime make sense. will do.

tomdee commented:
@mavenugo Do you mean this issue? https://github.com/docker/libnetwork/issues/603

If so, can you add it to a milestone so it can be fixed for 1.9

mavenugo commented:
@tiborvass @icecrime addressed the comments. PTAL.

@tomdee thanks for trying the code against calico driver. we fixed a few issues that you reported that are related to this PR. Pls give it a try.

mavenugo commented:
@tiborvass addressed the `IPAMConfig` and using MapOpts for AuxAddress. PTAL.

icecrime commented:
I still see labels in the current PR, am I missing something?

tomdee commented:
I'm trying out a new build from this branch but I can't get docker to start. Is this a problem that needs to be fixed or just an artifact of having run the previous code. If so, how do I sort it out?

Running with a clean consul, or without any datastore results in the following error
```
WARN[0000] couldn't roll back driver network on network none creation failure: failed to update store for object type *libnetwork.network: json: cannot unmarshal bool into Go value of type string 
DEBU[0000] Cleaning up old shm/mqueue mounts: start.    
FATA[0000] Error starting daemon: Error initializing network controller: Error creating default "null" network: failed to update store for object type *libnetwork.network: json: cannot unmarshal bool into Go value of type string 
```

And incidentally, if I run with the --cluster-store flag but the database can't be contacted I get the following which is a bit noisy.
```
ERRO[0000] getNetworkFromStore for nid 3953049d52e294a8ce1d0e0880ea5bf93c24f874d36bd5cf75765b2eab3e9d31 failed while trying to build sandbox for cleanup: could not find network 3953049d52e294a8ce1d0e0880ea5bf93c24f874d36bd5cf75765b2eab3e9d31: Get http://127.0.0.1:8500/v1/kv/docker/network/v1.0/network/3953049d52e294a8ce1d0e0880ea5bf93c24f874d36bd5cf75765b2eab3e9d31/?consistent=: dial tcp 127.0.0.1:8500: connection refused 
ERRO[0000] getNetworkFromStore for nid 43a788edab275cfadfe0a069b5a15a124a3d00f35e890c07f2accd4daa4b2fef failed while trying to build sandbox for cleanup: could not find network 43a788edab275cfadfe0a069b5a15a124a3d00f35e890c07f2accd4daa4b2fef: Get http://127.0.0.1:8500/v1/kv/docker/network/v1.0/network/43a788edab275cfadfe0a069b5a15a124a3d00f35e890c07f2accd4daa4b2fef/?consistent=: dial tcp 127.0.0.1:8500: connection refused 
ERRO[0000] getNetworkFromStore for nid 3953049d52e294a8ce1d0e0880ea5bf93c24f874d36bd5cf75765b2eab3e9d31 failed while trying to build sandbox for cleanup: could not find network 3953049d52e294a8ce1d0e0880ea5bf93c24f874d36bd5cf75765b2eab3e9d31: Get http://127.0.0.1:8500/v1/kv/docker/network/v1.0/network/3953049d52e294a8ce1d0e0880ea5bf93c24f874d36bd5cf75765b2eab3e9d31/?consistent=: dial tcp 127.0.0.1:8500: connection refused 
ERRO[0000] getNetworkFromStore for nid 43a788edab275cfadfe0a069b5a15a124a3d00f35e890c07f2accd4daa4b2fef failed while trying to build sandbox for cleanup: could not find network 43a788edab275cfadfe0a069b5a15a124a3d00f35e890c07f2accd4daa4b2fef: Get http://127.0.0.1:8500/v1/kv/docker/network/v1.0/network/43a788edab275cfadfe0a069b5a15a124a3d00f35e890c07f2accd4daa4b2fef/?consistent=: dial tcp 127.0.0.1:8500: connection refused 
WARN[0000] couldn't roll back driver network on network none creation failure: failed to update store for object type *libnetwork.network: json: cannot unmarshal bool into Go value of type string 
DEBU[0000] Cleaning up old shm/mqueue mounts: start.   
```

mavenugo commented:
@tomdee it is NOT a problem of the new code. it is the artifact of the data-model change, since you are trying intermediate versions in the PR. I will help you in the IRC to get past this.

tomdee commented:
OK, good to hear! I'm now unblocked by running `sudo rm /var/lib/docker/network/files/local-kv.db`

mavenugo commented:
@icecrime @tiborvass i hear your concerns. I will remove the labels/opts out of this PR and open a new one for that. Thanks for your help in reviewing this.

mavenugo commented:
@tomdee thanks for confirming. :+1:

tomdee commented:
OK, basic tests pass but I'm still seeing
* The gateway issue in Docker inspect output https://github.com/docker/docker/pull/16910#issuecomment-147544026
* And etcd not working (is that being tracked here - https://github.com/docker/docker/issues/16863)

I do see the improved error message when creating a network for a global driver without a backing datastore.
I couldn't easily test the /etc/hosts thing.

Where is the discussion about removing labels? I think it may have been lost with the rebase. I was looking forward to them :disappointed:

mavenugo commented:
@tiborvass @icecrime removed the labels from this PR. Will push another one for that.

@tomdee there was a valid concern that this PR is much bigger and requires more dedicated discussion around driver options. Hence, in order to unblock this PR, we removed it. Will add it as an independent PR.

tiborvass commented on api/client/network.go:
```diff
@@ -29,15 +33,37 @@ func (cli *DockerCli) CmdNetwork(args ...string) error {
+	cmd.Var(&flIpamSubnet, []string{"-subnet"}, "Subnet in CIDR format that represents a network segment")
+	cmd.Var(&flIpamIPRange, []string{"-ip-range"}, "allocate container ip from a sub-range")
+	cmd.Var(&flIpamGateway, []string{"-gateway"}, "ipv4 or ipv6 Gateway for the master subnet")
+	cmd.Var(flIpamAux, []string{"-aux-address"}, "Auxilary ipv4 or ipv6 addresses used by Network driver")
```
`Auxiliary IPv4 or IPv6` also: `network driver` lowercase

tiborvass commented on daemon/network/settings.go:
```diff
@@ -8,6 +8,20 @@ type Address struct {
+	Subnet     string            `json:"subnet,omitempty"`
+	IPRange    string            `json:"ip_range,omitempty"`
+	Gateway    string            `json:"gateway,omitempty"`
+	AuxAddress map[string]string `json:"auxilary_address,omitempty"`
```
`auxiliary_address` second `i` is missing

tiborvass commented on api/client/network.go:
```diff
@@ -186,6 +210,129 @@ func (cli *DockerCli) CmdNetworkInspect(args ...string) error {
+			}
+			d := iData[s]
+			d.IPRange = r
+			match = true
```
@mavenugo shouldn't it break here? If so, you could do:

```Go
if ok {
  if Data[s].IPRange != "" {
    ...
  }
  ...
  match = true
  break
}
```

mavenugo replied:
Since we support multiple subnet, ipranges, there cannot be multiple matches. Hence we should not break on the first match. All we are looking for is 1 exact match.

jfrazelle commented:
can you make sure that any tests you added will _not_ be run on lxc, so testRequires(c, NativeExecDriver)

icecrime commented:
It's green, LGTM.

jfrazelle commented:
ok ill fix the tests after if they need it :P

tiborvass commented:
LGTM

tiborvass commented:
Would be great to have bash completion for this before final release /cc @albers

thaJeztah commented:
@tiborvass think those are the changes in https://github.com/docker/docker/commit/cc6aece1fdefbc10638fe9e462a15608c6093115, 

thus `--subnet`, `--ip-range` and `--gateway` options in docker network command, and the  `--aux-address` option?

Also,

/cc @sdurrheimer for zsh completions :+1:

sdurrheimer commented:
@thaJeztah Will do as soon as possible

albers commented:
I'll take care of bash completion, thanks for pinging me.

thaJeztah commented:
Thanks both!

mavenugo commented:
@tomdee PTAL : https://github.com/docker/docker/pull/17046
Also, the etcd issue should have been resolved by now.
